### PR TITLE
Add the ability to auto-mock abstract classes

### PR DESCRIPTION
**Pull request by eliah-hecht-zocdoc**
Added the ability for AutoMock to mock abstract classes (previously it only did interfaces). Also added some test coverage, and did some refactoring for clarity.

### CODE REVIEW

File: src/Autofac.Extras.Moq/MoqRegistrationHandler.cs [hunk 1]
```diff
@@ -82,6 +79,31 @@ public IEnumerable<IComponentRegistration> RegistrationsFor
...
+                   !IsGenericType(typedService) &&
+                   !IsIStartable(typedService);
+        }
+
+        private static bool IsIStartable(IServiceWithType typedService)
+        {
+            return typeof (IStartable).IsAssignableFrom(typedService.ServiceType);
+        }
+
+        private static bool IsGenericType(IServiceWithType typedService)
```
Comment (tillig) [inline src/Autofac.Extras.Moq/MoqRegistrationHandler.cs]: Can we call this `IsUnsupportedGenericType`? If you read through the code it _appears from the name_ that you can't mock any generic type, which isn't true.

**Comment by tillig** [comment]
Very nice. I had one comment for a name clarity suggestion; if we can address that, I'll pull it and send it to MyGet so you can try it out.

File: src/Autofac.Extras.Moq/MoqRegistrationHandler.cs (hunk 1 above)
Comment (eliah-hecht-zocdoc) [inline src/Autofac.Extras.Moq/MoqRegistrationHandler.cs]: How about just `IsIEnumerable`?

Comment (tillig) [inline src/Autofac.Extras.Moq/MoqRegistrationHandler.cs]: Sure, that'd work, too.

**Comment by eliah-hecht-zocdoc** [comment]
OK, fixed. Thanks for the quick response.

**Comment by tillig** [comment]
Pushed to NuGet as v 3.1.0 because *I live dangerously*. :)

**Pull request merged**